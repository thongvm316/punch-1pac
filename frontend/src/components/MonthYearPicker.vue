<template>
  <div class="month-year-picker">
    <div class="month-year-picker-result" @click="togglePicker">
      {{ monthFormat }}
    </div>
    <div class="month-year-picker-content" v-show="isOpenMonthYearPicker" ref="result">
      <div class="month-year-picker-control">
        <span :class="{ 'active': (isOpenMonthYearPicker && isMonthPicker) }" @click="isMonthPicker = true">{{ $t('group.month') }}</span>
        <span :class="{ 'active': (isOpenMonthYearPicker && !isMonthPicker)}" @click="isMonthPicker = false">{{ $t('group.year') }}</span>
      </div>
      <datepicker
        :inline="true"
        :language="currentUser.language"
        :minimumView="isMonthPicker ? 'month' : 'year'"
        :maximumView="isMonthPicker ? 'month' : 'year'"
        :input-class="'form-input'"
        :calendar-class="'datepicker-calendar'"
        :wrapper-class="'datepicker'"
        v-model="month"
        @input="onInputDatepicker" />
    </div>
  </div>
</template>

<script>
import Datepicker from 'vuejs-datepicker'

export default {
  name: 'month-year-picker',

  data () {
    return {
      month: this.$moment().locale('en').format('YYYY-MM-DD'),
      isMonthPicker: true,
      isOpenMonthYearPicker: false
    }
  },

  components: {
    Datepicker
  },

  computed: {
    monthFormat () {
      return this.isMonthPicker ? this.$moment(this.month).format('LLL') : this.$moment(this.month).format('YYYY')
    }
  },

  methods: {
    togglePicker (event) {
      event.stopPropagation()
      this.isOpenMonthYearPicker = !this.isOpenMonthYearPicker
      const elResult = this.$refs.result

      if (!document.onclick) {
        document.onclick = (e) => {
          const target = e.target
          if ((target !== elResult.className) && (!elResult.contains(target))) {
            this.isOpenMonthYearPicker = false
          }
        }
      }
    },

    onInputDatepicker () {
      this.month = this.$moment(this.month).format('YYYY-MM-DD')
      this.$emit('input', {
        date: this.$moment(this.month).format('YYYY-MM-DD'),
        type: this.isMonthPicker ? 'month' : 'year'
      })
      this.isOpenMonthYearPicker = false
    }
  }
}
</script>
