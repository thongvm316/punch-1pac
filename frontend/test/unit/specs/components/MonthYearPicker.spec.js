import { shallowMount } from '@vue/test-utils'
import wrapperOps from '../../supports/wrapper'
import MonthYearPicker from '@/components/MonthYearPicker'

const togglePicker = jest.spyOn(MonthYearPicker.methods, 'togglePicker')
const onInputDatepicker = jest.spyOn(MonthYearPicker.methods, 'onInputDatepicker')

const localWrapperOps = {
  ...wrapperOps,
  Datepicker: true,
  data () {
    return {
      month: '2020-01-03',
      isMonthPicker: true,
      isOpenMonthYearPicker: true
    }
  }
}

describe('MonthYearPicker.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(MonthYearPicker, localWrapperOps)
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when MonthYearPicker was mounted', () => {
    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when methods', () => {
    it('should togglePicker was called', async () => {
      wrapper.find('.month-year-picker-result').trigger('click')
      await wrapper.vm.$nextTick()

      expect(togglePicker).toHaveBeenCalled()
    })

    it('should onInputDatepicker was called', async () => {
      const input = {
        date: '2020-01-03',
        type: 'month'
      }
      wrapper.vm.onInputDatepicker()
      await wrapper.vm.$nextTick()

      expect(wrapper.emitted('input')).toBeTruthy()
      expect(wrapper.emitted('input')[0]).toEqual([input])
    })
  })
})
