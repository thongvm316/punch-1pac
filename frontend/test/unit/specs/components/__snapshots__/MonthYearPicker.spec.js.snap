// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthYearPicker.vue when MonthYearPicker was mounted should render correctly 1`] = `
<div class="month-year-picker">
  <div class="month-year-picker-result">
    Jan 2020
  </div>
  <div class="month-year-picker-content">
    <div class="month-year-picker-control"><span class="active">Month</span> <span class="">Year</span></div>
    <div class="vdp-datepicker datepicker">
      <div class="">
        <!----> <input type="hidden" readonly="readonly" class="form-input" value="03 Jan 2020">
        <!---->
      </div>
      <!---->
      <div class="datepicker-calendar vdp-datepicker__calendar" style="position: static;">
        <header><span class="prev">&lt;</span> <span class="">2020</span> <span class="next">&gt;</span></header> <span track-by="timestamp" class="cell month selected">January</span><span track-by="timestamp" class="cell month">February</span><span track-by="timestamp" class="cell month">March</span><span track-by="timestamp" class="cell month">April</span><span track-by="timestamp" class="cell month">May</span><span track-by="timestamp" class="cell month">June</span><span track-by="timestamp" class="cell month">July</span><span track-by="timestamp" class="cell month">August</span><span track-by="timestamp" class="cell month">September</span><span track-by="timestamp" class="cell month">October</span><span track-by="timestamp" class="cell month">November</span><span track-by="timestamp" class="cell month">December</span>
      </div>
      <!---->
    </div>
  </div>
</div>
`;
